Style/IfUnlessModifier does not trigger if the body is another conditional

amuino opened the pull request:
This is a proposal to avoid adding an offense when the body of the conditional is another conditional (originally targeting the ternary operator), since the resulting one-liner is a lot harder to read.

When the content of an `if` sentence is a conditional (specially with the
ternary operator), the single line version is much harder to read.

Before, conditions like
```
if running?
  tired? ? "please stop" : "keep going"
end
```

become

```
tired? ? "please stop" : "keep going" if running?
```

With this PR, the original remains unchanged.

rrosenblum commented:
The branch is going to need to be rebased. Aside from that, the change looks good to me.

bbatsov commented:
My concern is that it might not be apparent to some people why the cop would ignore such cases. This is why I hate special cases in general - it's really subjective whether something should be treated differently or not. An explicit option would make this better IMO.

amuino commented:
Rebased.

Regarding this being an _special case_, I would rather focus on the conversation on whether 

```
tired? ? "please stop" : "keep going" if running?
``` 

is the kind of code that we want to recommend or not, and if a modification to [Ruby style guide is needed](https://github.com/bbatsov/ruby-style-guide#if-as-a-modifier).

Our team's opinion is that ternary operator followed by `if/unless` is hard to read and should be avoided, but we haven't checked if there is a general consensus. I can work on adding a toggle to enable them if needed, but would rather keep the setting off by default.

bbatsov commented:
On second thought there should probably be a different cop flagging code like:

```
tired? ? "please stop" : "keep going" if running?
```

and other similar constructs.

bbatsov commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # Change log
 
 ## master (unreleased)
+* `Style/IfUnlessModifier` does not trigger if the body is another conditional. ([@amuino][])
```
This should be in changes.

bbatsov commented on lib/rubocop/cop/style/if_unless_modifier.rb:
```diff
@@ -26,6 +26,7 @@ def on_if(node)
           return if if_else?(node)
           return if node.chained?
           return unless fit_within_line_as_modifier_form?(node)
+          return if node.descendants.any? { |d| d.type == :if }
```
This looks wrong to me - this should apply only to cases where the body is just some form of an `if` expression (meaning you should have just a single descendent).

amuino replied:
I think in practice there will be a single descendant since we already discard bodies with several statements ([per this existing test](https://github.com/amuino/rubocop/blob/if-unless-modifier-with-nested-conditional/spec/rubocop/cop/style/if_unless_modifier_spec.rb#L43-L54)) , so I'll change that.

bbatsov replied:
True.

amuino replied:
Updated to check only direct `then`and `else` children.

bbatsov commented on spec/rubocop/cop/style/if_unless_modifier_spec.rb:
```diff
@@ -310,4 +310,17 @@
+
+  context 'if-end with conditional as body' do
+    let(:source) do
+      ['if condition',
```
There should also be tests showing that a body with several expressions, one of which is `if` is not affected.

amuino replied:
Happy to add it, but isn't that already covered by [this more general case](https://github.com/amuino/rubocop/blob/if-unless-modifier-with-nested-conditional/spec/rubocop/cop/style/if_unless_modifier_spec.rb#L43-L54)?

bbatsov replied:
Good point.

amuino commented:
I think I've addresses all comments to this PR except for the additional cop for
```
tired? ? "please stop" : "keep going" if running?
```

I can work on that as a different PR. What do you think?

bbatsov commented on lib/rubocop/cop/style/if_unless_modifier.rb:
```diff
@@ -64,6 +65,13 @@ def autocorrect(node)
+
+        # returns false if the then or else children are conditionals
+        def nested_conditional?(node)
+          node.children[1,2].any? { |child| child && child.type == :if }
```
1, 2

bbatsov commented:
> I can work on that as a different PR. What do you think?
>  Hide all checks

That'd be great. You should also fix the failing build and squash the 3 commits together.

bbatsov commented on lib/rubocop/cop/style/if_unless_modifier.rb:
```diff
@@ -64,6 +65,13 @@ def autocorrect(node)
           ->(corrector) { corrector.replace(node.source_range, oneline) }
         end
+
+        protected
```
Why protected?

amuino replied:
I'm used to make methods `protected` when I don't want them on the class API but might be useful for subclasses, not really attached to that (I think is what remains of a past as a java developer), let me know what you would prefer.

bbatsov replied:
`protected` has a different meaning in Ruby. Just make this `private`.

amuino commented:
Run rubocop to fix the style issue and sqashed.

amuino commented:
`protected` → `private`

bbatsov commented:
:+1:
